### PR TITLE
DA-3331: Adding org service related service methods

### PR DESCRIPTION
**Pull request by ajinkyan83**
Jira Ticket: https://jira.linuxfoundation.org/browse/DA-3331

**Details:**
This PR adds related code for org service i.e. https://api-gw.dev.platform.linuxfoundation.org/organization-service/v1/api-docs.

**APIs covered:**
- SearchOrganization - https://api-gw.dev.platform.linuxfoundation.org/organization-service/v1/orgs/search?name=linux
- LookupOrganization -https://api-gw.dev.platform.linuxfoundation.org/organization-service/v1/lookup?name=linux

**Testing done:**
Initially verified the org service dev endpoint request & response using Postman.
Testing done locally using dev platform endpoints and DA test environment.

**Unit Test Case Coverage**
```
$ go test -v
=== RUN   TestLookupOrganization
--- PASS: TestLookupOrganization (0.00s)
=== RUN   TestSearchOrganization
--- PASS: TestSearchOrganization (0.00s)
=== RUN   TestSearchOrganizationSpecialChars
--- PASS: TestSearchOrganizationSpecialChars (0.00s)
PASS
ok      github.com/LF-Engineering/dev-analytics-libraries/orgs  0.278s
```

### CODE REVIEW

File: orgs/organization.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+	if name == "" {
+		log.Println("SearchOrganization: name is empty")
+		return nil
+	}
+	token, err := o.auth0Client.ValidateToken(o.Environment)
+	if err != nil {
+		log.Println(err)
+	}
+	headers := make(map[string]string, 0)
+	headers["Content-type"] = "application/json"
```
Comment (code-sleuth) [inline orgs/organization.go:43]: the `httpClient.Request` method already adds a content type header. lets remove this

File: orgs/organization.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+	headers["Authorization"] = fmt.Sprintf("%s %s", "Bearer", token)
+
+	if offset == "" {
+		offset = "0"
+	}
+	if pageSize == "" {
+		pageSize = "100"
+	}
+	endpoint := o.OrgBaseURL + "/orgs/search?name=" + name + "&pageSize=" + pageSize + "&offset=" + offset
```
Comment (code-sleuth) [inline orgs/organization.go:52]: url escape the org name

File: orgs/organization.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+	if name == "" {
+		log.Println("LookupOrganization: name is empty")
+		return nil
+	}
+	token, err := o.auth0Client.ValidateToken(o.Environment)
+	if err != nil {
+		log.Println(err)
+	}
+	headers := make(map[string]string, 0)
+	headers["Content-type"] = "application/json"
```
Comment (code-sleuth) [inline orgs/organization.go:78]: the httpClient.Request method already adds a content type header. lets remove this

File: orgs/organization.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+	headers["Authorization"] = fmt.Sprintf("%s %s", "Bearer", token)
+
+	endpoint := o.OrgBaseURL + "/lookup?name=" + name
```
Comment (code-sleuth) [inline orgs/organization.go:81]: url escape the org name, `url.PathEscape(name)`

File: orgs/organization.go [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+	if err != nil {
+		log.Println("LookupOrganization: failed to unmarshal LookupOrganizationResponse: ", err)
+		return nil
+	}
+	return &response
+}
+
+// NewOrganizationsClient consumes
+// orgBaseURL, esCacheUrl, esCacheUsername, esCachePassword, esCacheIndex, env, authGrantType, authClientID, authClientSecret, authAudience, authURL
+func NewOrganizationsClient(orgBaseURL, esCacheURL, esCacheUsername,
```
Comment (enkhalifapro) [inline orgs/organization.go:98]: Just NewClient since the package called organization

File: orgs/organization.go [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+		log.Println("SearchOrganization: failed to unmarshal SearchOrganizationResponse: ", err)
+		return nil
+	}
+	return &response
+}
+
+// LookupOrganization ...
+func (o *Org) LookupOrganization(name string) *Organization {
+	if name == "" {
+		log.Println("LookupOrganization: name is empty")
```
Comment (enkhalifapro) [inline orgs/organization.go:70]: Return error

File: orgs/organization.go [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+		return nil
+	}
+	token, err := o.auth0Client.ValidateToken(o.Environment)
+	if err != nil {
+		log.Println(err)
```
Comment (enkhalifapro) [inline orgs/organization.go:75]: return error to be handled in consumer side

File: orgs/organization.go (hunk 4 above)
Comment (code-sleuth) [reply orgs/organization.go:81]: this is because the name could have special characters

File: orgs/organization.go [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+	}
+	headers := make(map[string]string, 0)
+	headers["Content-type"] = "application/json"
+	headers["Authorization"] = fmt.Sprintf("%s %s", "Bearer", token)
+
+	endpoint := o.OrgBaseURL + "/lookup?name=" + name
+	_, res, err := o.httpClient.Request(strings.TrimSpace(endpoint), "GET", headers, nil, nil)
+	if err != nil {
+		log.Println("LookupOrganization: Could not get the organization: ", err)
```
Comment (enkhalifapro) [inline orgs/organization.go:84]: Same comment

File: orgs/organization.go [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+		return nil
+	}
+	var response Organization
+	err = json.Unmarshal(res, &response)
+	if err != nil {
+		log.Println("LookupOrganization: failed to unmarshal LookupOrganizationResponse: ", err)
```
Comment (enkhalifapro) [inline orgs/organization.go:90]: same comment

File: orgs/organization.go [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+	}
+
+	org.esClient = esClientProvider
+	org.httpClient = httpClientProvider
+	org.auth0Client = auth0ClientProvider
+
+	return org, nil
+}
+
+func buildServices(o *Org) (httpClientProvider *http.ClientProvider, esClientProvider *elastic.ClientProvider, auth0ClientProvider *auth0.ClientProvider, err error) {
```
Comment (enkhalifapro) [inline orgs/organization.go:126]: Based on Golang codeReview doc, named output params is not good approach, remove the names and return explicit vars

**Comment by code-sleuth** [comment]
Generally, I think this should be a client for the ProjectsService. I know we are only covering orgs for this task but if we transform this to be the projects service client, it will create room for other endpoints we might want to consume in future without duplicating functionality to the same service.

File: orgs/organization.go (hunk 1 above)
Comment (ajinkyan83) [reply orgs/organization.go]: done!

File: orgs/organization.go (hunk 3 above)
Comment (ajinkyan83) [reply orgs/organization.go]: done!

File: orgs/organization.go (hunk 5 above)
Comment (ajinkyan83) [reply orgs/organization.go]: done!

File: orgs/organization.go (hunk 6 above)
Comment (ajinkyan83) [reply orgs/organization.go]: done!

File: orgs/organization.go (hunk 7 above)
Comment (ajinkyan83) [reply orgs/organization.go:77]: done!

File: orgs/organization.go (hunk 8 above)
Comment (ajinkyan83) [reply orgs/organization.go:86]: done!

File: orgs/organization.go (hunk 10 above)
Comment (ajinkyan83) [reply orgs/organization.go]: done!

File: orgs/organization.go (hunk 9 above)
Comment (ajinkyan83) [reply orgs/organization.go:92]: done!

**Review by code-sleuth — APPROVED** [review]
/lgtm

**Review by enkhalifapro — CHANGES_REQUESTED** [review]
Cover the new functions with required unit tests

**Comment by ajinkyan83** [comment]
> Cover the new functions with required unit tests

Done. 
Mocks generated using `https://github.com/vektra/mockery`

**Review by enkhalifapro — APPROVED** [review]

**Pull request merged**